python/lsst/pipe/base/graphBuilder.py

```diff
@@ -564,108 +586,80 @@ def fillDatasetRefs(self, registry, collections, run, *, skipExisting=True):
+            for datasetType, refs in itertools.chain(self.initIntermediates.items(),
+                                                     self.initOutputs.items(),
+                                                     self.intermediates.items(),
+                                                     self.outputs.items()):
+                _LOG.debug("Resolving %d datasets for intermediate and/or output dataset %s.",
+                           len(refs), datasetType.name)
+                for dataId, unresolvedRef in refs.items():
+                    # TODO: we could easily support per-DatasetType skipExisting
+                    # and I could imagine that being useful - it's probably
+                    # required in order to support writing initOutputs before
+                    # QuantumGraph generation.
                     ref = registry.findDataset(datasetType=datasetType, dataId=dataId, collections=run)
-                else:
-                    ref = None
-                if ref is None:
-                    ref = DatasetRef(datasetType, dataId)
-                elif not skipExisting:
-                    raise OutputExistsError(f"Output dataset {datasetType.name} already exists in "
-                                            f"output RUN collection '{run}' with data ID {dataId}.")
-                scaffolding.refs.append(ref)
-        # Prerequisite dataset lookups are deferred until fillQuanta.
-
-    def fillQuanta(self, registry, collections, *, skipExisting=True):
-        """Define quanta for each task by splitting up the datasets associated
-        with each task data ID.
-
-        This method populates `_TaskScaffolding.quanta`.
-
-        Parameters
-        ----------
-        registry : `lsst.daf.butler.Registry`
-            Registry for the data repository; used for all data ID queries.
-        collections : `lsst.daf.butler.CollectionSearch`
-            Object representing the collections to search for input datasets.
-        skipExisting : `bool`, optional
-            If `True` (default), a Quantum is not created if all its outputs
-            already exist.
-        """
+                    if ref is None:
+                        ref = unresolvedRef
+                    elif not skipExisting:
+                        raise OutputExistsError(f"Output dataset {datasetType.name} already exists in "
+                                                f"output RUN collection '{run}' with data ID {dataId}.")
+                    refs[dataId] = ref
+        # Look up input and initInput datasets in the input collection(s).
+        for datasetType, refs in itertools.chain(self.initInputs.items(), self.inputs.items()):
+            _LOG.debug("Resolving %d datasets for input dataset %s.", len(refs), datasetType.name)
+            for dataId in refs:
+                refs[dataId] = registry.findDataset(datasetType, dataId=dataId, collections=collections)
+            assert not any(ref is None for ref in refs.values())
```

Comment:
bare asserts can be hard for debug. Its up to you to decide if it is likely enough to happen to make a richer message worth it.

Reply:
Yeah, this one is probably worth changing to a `RuntimeError` raise, as it's guarding against both daf_butler logic (which is probably too "far away" for this assert), _and_ that these datasets haven't been deleted by some other Registry client since the previous query.